src/models/event.js

```diff
@@ -327,17 +347,50 @@ utils.extend(module.exports.MatrixEvent.prototype, {
             );
         }
 
-        try {
-            crypto.decryptEvent(this);
-        } catch (e) {
-            console.warn(
-                `Error decrypting event (id=${this.getId()}): ${e}`,
+        if (this._decrypting) {
+            console.log(
+                `Event ${this.getId()} already being decrypted; queueing a retry`,
             );
+            this._retryDecryption = true;
+            return;
+        }
+
+        this._decrypting = true;
+
+        this._doDecryption(crypto).finally(() => {
+            this._decrypting = false;
+            this._retryDecryption = false;
```

Comment:
I guess my point is more that it's unclear that when you call `event.attemptDecryption` a 2nd time, that it will sometimes maybe attempt a 2nd decryption iff we previously failed with a `DecryptionError`.

It seems backwards for the decryption function to be dictating that "no really, you can't retry decrypting this" when surely it should be up to the caller to decide this. My concern is that there will be a case whereby decryption fails without a `DecryptionError` and the "obvious" fix of adjusting internal state followed by `attemptDecryption` again will just not work because of this function's implementation. Can you allay my concerns?